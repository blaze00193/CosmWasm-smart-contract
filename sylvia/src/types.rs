use cosmwasm_std::{Deps, DepsMut, Env, MessageInfo};

pub struct ReplyCtx<'a> {
    pub deps: DepsMut<'a>,
    pub env: Env,
}

pub struct MigrateCtx<'a> {
    pub deps: DepsMut<'a>,
    pub env: Env,
}

pub struct ExecCtx<'a> {
    pub deps: DepsMut<'a>,
    pub env: Env,
    pub info: MessageInfo,
}

pub type InstantiateCtx<'a> = ExecCtx<'a>;

pub struct QueryCtx<'a> {
    pub deps: Deps<'a>,
    pub env: Env,
}

impl ExecCtx<'_> {
    pub fn branch(&'_ mut self) -> ExecCtx<'_> {
        ExecCtx {
            deps: self.deps.branch(),
            env: self.env.clone(),
            info: self.info.clone(),
        }
    }
}

impl<'a> From<(DepsMut<'a>, Env)> for ReplyCtx<'a> {
    fn from((deps, env): (DepsMut<'a>, Env)) -> Self {
        Self { deps, env }
    }
}

impl<'a> From<(DepsMut<'a>, Env)> for MigrateCtx<'a> {
    fn from((deps, env): (DepsMut<'a>, Env)) -> Self {
        Self { deps, env }
    }
}

impl<'a> From<(DepsMut<'a>, Env, MessageInfo)> for ExecCtx<'a> {
    fn from((deps, env, info): (DepsMut<'a>, Env, MessageInfo)) -> Self {
        Self { deps, env, info }
    }
}

impl<'a> From<(Deps<'a>, Env)> for QueryCtx<'a> {
    fn from((deps, env): (Deps<'a>, Env)) -> Self {
        Self { deps, env }
    }
}
