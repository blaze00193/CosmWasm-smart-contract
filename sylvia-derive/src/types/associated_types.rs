use proc_macro2::{Ident, TokenStream};
use quote::quote;
use syn::{parse_quote, ItemTrait, TraitItem, TraitItemType, Type, WhereClause, WherePredicate};

pub const ERROR_TYPE: &str = "Error";
pub const EXEC_TYPE: &str = "ExecC";
pub const QUERY_TYPE: &str = "QueryC";

/// Wrapper around associated types in parsed from a trait.
#[derive(Default)]
pub struct AssociatedTypes<'a>(Vec<&'a TraitItemType>);

impl<'a> AssociatedTypes<'a> {
    pub fn new(source: &'a ItemTrait) -> Self {
        let associated_types: Vec<_> = source
            .items
            .iter()
            .filter_map(|item| match item {
                TraitItem::Type(ty) => Some(ty),
                _ => None,
            })
            .collect();

        Self(associated_types)
    }

    /// Returns [Iterator] over underlying [TraitItemType]s mapped to [Ident]s.
    pub fn as_names(&self) -> impl Iterator<Item = &Ident> {
        self.0.iter().map(|associated| &associated.ident)
    }

    /// Returns [Iterator] over underlying [TraitItemType]s without the `Error` type.
    /// Used for generating generics for generated types.
    pub fn without_error(&self) -> impl Iterator<Item = &TraitItemType> {
        self.0
            .iter()
            .filter(|associated| associated.ident != ERROR_TYPE)
            .cloned()
    }

    /// Returns [WherePredicate] from underlying [TraitItemType]s.
    pub fn as_where_predicates(&self) -> Vec<WherePredicate> {
        self.without_error()
            .map(|associated| {
                let name = &associated.ident;
                let colon = &associated.colon_token;
                let bound = &associated.bounds;
                parse_quote! { #name #colon #bound }
            })
            .collect()
    }

    /// Returns [WhereClause] from underlying [TraitItemType]s.
    pub fn as_where_clause(&self) -> Option<WhereClause> {
        let predicates = self.as_where_predicates();
        if !predicates.is_empty() {
            parse_quote! { where #(#predicates),* }
        } else {
            None
        }
    }

    /// Returns [WhereClause] from underlying [TraitItemType]s.
    pub fn emit_contract_predicate(&self, trait_name: &Ident) -> TokenStream {
        let predicate = quote! { ContractT: #trait_name };
        if self.0.is_empty() {
            return predicate;
        }

        let bounds = self.without_error().map(|associated| {
            let name = &associated.ident;
            quote! { #name = #name }
        });

        quote! {
            #predicate < #(#bounds,)* >
        }
    }

    /// Returns `Some(Type)` if `type_name` is present in the underlying associated types.
    /// Returns `None` if it is not.
    pub fn emit_contract_custom_type_accessor(
        &self,
        trait_name: &Ident,
        type_name: &str,
    ) -> Option<Type> {
        match self
            .as_names()
            .find(|name| name.to_string().as_str() == type_name)
        {
            Some(name) => {
                let type_name = Ident::new(type_name, name.span());
                Some(parse_quote! { <ContractT as #trait_name>:: #type_name})
            }
            None => None,
        }
    }
}

/// Trait defining convertion to [Ident] and [WherePredicate].
pub trait ItemType {
    fn as_name(&self) -> &Ident;
    fn as_where_predicate(&self) -> WherePredicate;
}

impl ItemType for TraitItemType {
    fn as_name(&self) -> &Ident {
        &self.ident
    }

    fn as_where_predicate(&self) -> WherePredicate {
        let name = &self.ident;
        let colon = &self.colon_token;
        let bound = &self.bounds;
        parse_quote! { #name #colon #bound }
    }
}

/// Trait generating associated types.
pub trait EmitAssociated {
    fn emit_declaration(&self) -> Vec<TokenStream>;
    fn emit_implementation(&self) -> Vec<TokenStream>;
}

impl EmitAssociated for WhereClause {
    fn emit_declaration(&self) -> Vec<TokenStream> {
        self.predicates
            .iter()
            .filter_map(|predicate| match predicate {
                WherePredicate::Type(predicate) => {
                    let bounded_ty = &predicate.bounded_ty;
                    let bounds = &predicate.bounds;
                    let lifetimes = &predicate.lifetimes.as_ref().map(|lf| {
                        let lf = &lf.lifetimes;
                        quote! { < #lf > }
                    });
                    Some(quote! { type #bounded_ty #lifetimes: #bounds; })
                }
                _ => None,
            })
            .collect()
    }

    fn emit_implementation(&self) -> Vec<TokenStream> {
        self.predicates
            .iter()
            .filter_map(|predicate| match predicate {
                WherePredicate::Type(predicate) => {
                    let bounded_ty = &predicate.bounded_ty;
                    let lifetimes = &predicate.lifetimes.as_ref().map(|lf| {
                        let lf = &lf.lifetimes;
                        quote! { < #lf > }
                    });
                    Some(quote! { type #bounded_ty #lifetimes = #bounded_ty; })
                }
                _ => None,
            })
            .collect()
    }
}
