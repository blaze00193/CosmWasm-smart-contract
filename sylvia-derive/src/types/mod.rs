pub mod associated_types;
pub mod interfaces;
pub mod msg_field;
pub mod msg_type;
pub mod msg_variant;
