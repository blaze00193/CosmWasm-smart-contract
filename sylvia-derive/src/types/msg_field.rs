use crate::parser::check_generics::{CheckGenerics, GetPath};
use crate::strip_self_path::StripSelfPath;
use proc_macro2::TokenStream;
use proc_macro_error::emit_error;
use quote::quote;
use syn::fold::Fold;
use syn::spanned::Spanned;
use syn::visit::Visit;
use syn::{Attribute, Ident, Pat, PatType, Type};

/// Representation of single message variant field
#[derive(Debug)]
pub struct MsgField<'a> {
    name: &'a Ident,
    ty: &'a Type,
    stripped_ty: Type,
    attrs: &'a Vec<Attribute>,
}

impl<'a> MsgField<'a> {
    /// Creates new field from trait method argument
    pub fn new<Generic>(
        item: &'a PatType,
        generics_checker: &mut CheckGenerics<Generic>,
    ) -> Option<MsgField<'a>>
    where
        Generic: GetPath + PartialEq,
    {
        let name = match &*item.pat {
            Pat::Ident(p) => Some(&p.ident),
            pat => {
                // TODO: Support pattern arguments, when decorated with argument with item
                // name
                //
                // Eg.
                //
                // ```
                // fn exec_foo(&self, ctx: Ctx, #[sv::msg(name=metadata)] SomeData { addr, sender }: SomeData);
                // ```
                //
                // should expand to enum variant:
                //
                // ```
                // ExecFoo {
                //   metadata: SomeDaa
                // }
                // ```
                emit_error!(pat.span(), "Expected argument name, pattern occurred");
                None
            }
        }?;

        let ty = &item.ty;
        let stripped_ty = StripSelfPath.fold_type((*item.ty).clone());
        let attrs = &item.attrs;
        generics_checker.visit_type(&stripped_ty);

        Some(Self {
            name,
            ty,
            stripped_ty,
            attrs,
        })
    }

    /// Emits message field
    pub fn emit(&self) -> TokenStream {
        let Self {
            name,
            stripped_ty,
            attrs,
            ..
        } = self;

        quote! {
            #(#attrs)*
            #name: #stripped_ty
        }
    }

    /// Emits struct field
    pub fn emit_pub(&self) -> TokenStream {
        let Self {
            name,
            stripped_ty,
            attrs,
            ..
        } = self;

        quote! {
            #(#attrs)*
            pub #name: #stripped_ty
        }
    }

    /// Emits method field
    pub fn emit_method_field(&self) -> TokenStream {
        let Self {
            name, stripped_ty, ..
        } = self;

        quote! {
            #name: #stripped_ty
        }
    }

    pub fn emit_method_field_folded(&self) -> TokenStream {
        let Self { name, ty, .. } = self;

        quote! {
            #name: #ty
        }
    }

    pub fn name(&self) -> &'a Ident {
        self.name
    }
}
