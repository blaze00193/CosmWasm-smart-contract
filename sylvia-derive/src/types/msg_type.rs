use proc_macro2::TokenStream;
use proc_macro_error::emit_error;
use quote::quote;
use syn::fold::Fold;
use syn::{parse_quote, GenericParam, Ident, Type};

use crate::crate_module;
use crate::fold::StripSelfPath;
use crate::parser::attributes::msg::MsgType;
use crate::parser::Customs;

impl MsgType {
    pub fn emit_ctx_type(self, query_type: &Type) -> TokenStream {
        use MsgType::*;

        let sylvia = crate_module();

        match self {
            Exec | Instantiate => quote! {
                (#sylvia ::cw_std::DepsMut< #query_type >, #sylvia ::cw_std::Env, #sylvia ::cw_std::MessageInfo)
            },
            Migrate | Reply | Sudo => quote! {
                (#sylvia ::cw_std::DepsMut< #query_type >, #sylvia ::cw_std::Env)
            },
            Query => quote! {
                (#sylvia ::cw_std::Deps< #query_type >, #sylvia ::cw_std::Env)
            },
        }
    }

    pub fn emit_ctx_dispatch_values(self, customs: &Customs) -> TokenStream {
        use MsgType::*;

        match (self, customs.has_query) {
            (Exec, true) => quote! {
                (ctx.0.into_empty(), ctx.1, ctx.2)
            },
            (Query, true) | (Sudo, true) => quote! {
                (ctx.0.into_empty(), ctx.1)
            },
            _ => quote! { ctx },
        }
    }

    pub fn emit_ctx_params(self, query_type: &Type) -> TokenStream {
        use MsgType::*;

        let sylvia = crate_module();

        match self {
            Exec | Instantiate => quote! {
                deps: #sylvia ::cw_std::DepsMut< #query_type>, env: #sylvia ::cw_std::Env, info: #sylvia ::cw_std::MessageInfo
            },
            Migrate | Reply | Sudo => quote! {
                deps: #sylvia ::cw_std::DepsMut< #query_type>, env: #sylvia ::cw_std::Env
            },
            Query => quote! {
                deps: #sylvia ::cw_std::Deps< #query_type>, env: #sylvia ::cw_std::Env
            },
        }
    }

    pub fn emit_ep_name(self) -> Ident {
        match self {
            Self::Exec => parse_quote! { execute },
            Self::Instantiate => parse_quote! { instantiate },
            Self::Migrate => parse_quote! { migrate },
            Self::Sudo => parse_quote! { sudo },
            Self::Reply => parse_quote! { reply },
            Self::Query => parse_quote! { query },
        }
    }

    pub fn emit_ctx_values(self) -> TokenStream {
        use MsgType::*;

        match self {
            Exec | Instantiate => quote! { deps, env, info },
            Migrate | Reply | Query | Sudo => quote! { deps, env },
        }
    }

    /// Emits type which should be returned by dispatch function for this kind of message
    pub fn emit_result_type(self, msg_type: &Type, err_type: &Type) -> TokenStream {
        use MsgType::*;

        let sylvia = crate_module();

        match self {
            Exec | Instantiate | Migrate | Reply | Sudo => {
                quote! {
                    std::result::Result< #sylvia:: cw_std::Response <#msg_type>, #err_type>
                }
            }
            Query => quote! {
                std::result::Result<#sylvia ::cw_std::Binary, #err_type>
            },
        }
    }

    pub fn emit_msg_wrapper_name(&self) -> Ident {
        match self {
            MsgType::Exec => parse_quote! { ContractExecMsg },
            MsgType::Query => parse_quote! { ContractQueryMsg },
            MsgType::Sudo => parse_quote! { ContractSudoMsg },
            _ => self.emit_msg_name(),
        }
    }

    pub fn emit_msg_name(&self) -> Ident {
        match self {
            MsgType::Exec => parse_quote! { ExecMsg },
            MsgType::Query => parse_quote! { QueryMsg },
            MsgType::Instantiate => parse_quote! { InstantiateMsg },
            MsgType::Migrate => parse_quote! { MigrateMsg },
            MsgType::Reply => parse_quote! { ReplyMsg },
            MsgType::Sudo => parse_quote! { SudoMsg },
        }
    }

    pub fn as_accessor_wrapper_name(&self) -> Type {
        match self {
            MsgType::Exec => parse_quote! { ContractExec },
            MsgType::Query => parse_quote! { ContractQuery },
            MsgType::Sudo => parse_quote! { ContractSudo },
            _ => self.as_accessor_name(),
        }
    }

    pub fn as_accessor_name(&self) -> Type {
        match self {
            MsgType::Instantiate => parse_quote! { Instantiate },
            MsgType::Exec => parse_quote! { Exec },
            MsgType::Query => parse_quote! { Query },
            MsgType::Migrate => parse_quote! { Migrate },
            MsgType::Sudo => parse_quote! { Sudo },
            MsgType::Reply => parse_quote! { Reply },
        }
    }

    pub fn emit_phantom_variant(&self, generics: &[&GenericParam]) -> TokenStream {
        match self {
            _ if generics.is_empty() => quote! {},
            MsgType::Query => quote! {
                #[serde(skip)]
                #[returns(( #(#generics,)* ))]
                _Phantom(std::marker::PhantomData<( #(#generics,)* )>),
            },
            _ => quote! {
                #[serde(skip)]
                _Phantom(std::marker::PhantomData<( #(#generics,)* )>),
            },
        }
    }

    pub fn emit_derive_call(&self) -> TokenStream {
        let sylvia = crate_module();
        let cw_schema = quote! { #sylvia:: cw_schema }.to_string();
        let schemars = quote! { #sylvia:: cw_schema::schemars }.to_string();
        let serde = quote! { #sylvia:: serde }.to_string();

        match self {
            MsgType::Query => quote! {
                #[derive(#sylvia ::serde::Serialize, #sylvia ::serde::Deserialize, Clone, Debug, PartialEq, #sylvia ::schemars::JsonSchema, #sylvia:: cw_schema::QueryResponses)]
                #[schemars(crate = #schemars )]
                #[query_responses(crate = #cw_schema )]
                #[serde(crate = #serde )]
            },
            _ => quote! {
                #[derive(#sylvia ::serde::Serialize, #sylvia ::serde::Deserialize, Clone, Debug, PartialEq, #sylvia ::schemars::JsonSchema)]
                #[schemars(crate = #schemars )]
                #[serde(crate = #serde )]
            },
        }
    }

    pub fn emit_dispatch_leg(&self, function_name: &Ident, args: &Vec<Ident>) -> TokenStream {
        use MsgType::*;
        let sylvia = crate_module();

        match self {
            Exec | Sudo => quote! {
                contract.#function_name(Into::into(ctx), #(#args),*).map_err(Into::into)
            },
            Query => quote! {
                #sylvia ::cw_std::to_json_binary(&contract.#function_name(Into::into(ctx), #(#args),*)?).map_err(Into::into)
            },
            Instantiate | Migrate | Reply => {
                emit_error!(function_name.span(), "Internal Error";
                note = "Dispatch leg should be called only for `Enum` type messages.");
                quote! {}
            }
        }
    }

    pub fn emit_returns_attribute(&self, return_type: &Option<Type>) -> TokenStream {
        match (self, return_type) {
            (MsgType::Query, Some(return_type)) => {
                let stripped_return_type = StripSelfPath.fold_type(return_type.clone());
                quote! { #[returns(#stripped_return_type)] }
            }
            _ => quote! {},
        }
    }
}
