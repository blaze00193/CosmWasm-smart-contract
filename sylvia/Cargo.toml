[package]
name = "sylvia"
version = "0.3.2"
authors = ["Bart≈Çomiej Kuras <bart.k@confio.gmbh>"]
edition = "2021"
description = "Framework for generating trait based CosmWasm smart contracts"
license = "Apache-2.0"
repository = "https://github.com/CosmWasm/sylvia"
homepage = "https://cosmwasm.com"
keywords = ["cosmwasm", "blockchain", "confio"]
categories = ["api-bindings", "wasm"]

[features]
mt = ["sylvia-derive/mt", "dep:cw-multi-test", "dep:anyhow", "dep:derivative"]
no_entry_points = ["sylvia-derive/no_entry_points"]

[dependencies]
sylvia-derive = { version = "0.3.2", path = "../sylvia-derive" }
cosmwasm-std = { version = "1.2", features = ["staking"] }
cosmwasm-schema = "1.2"
schemars = "0.8"
serde = { version = "1.0", default-features = false, features = ["derive"] }
serde-cw-value = "0.7.0"
serde-json-wasm = "0.5.0"
konst = "0.3"
cw-multi-test = { version = "0.16.2", optional = true }
anyhow = { version = "1.0", optional = true }
derivative = { version = "2.2.0", optional = true }

[dev-dependencies]
anyhow = "1.0"
cw-storage-plus = "1.0.1"
cw-multi-test = "0.16.2"
thiserror = "1.0.38"
