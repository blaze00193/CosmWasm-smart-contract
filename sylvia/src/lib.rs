//! Framework for creating CosmWasm Smart Contract with high-level abstraction layer
//!
//! Most of implementation lies in `sylvia-derive` crate which is reexported here

pub mod into_response;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub mod multitest;
pub mod types;
pub mod utils;

#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use anyhow;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use cw_multi_test;
#[cfg_attr(docsrs, doc(cfg(feature = "cosmwasm_1_2")))]
#[cfg(feature = "cosmwasm_1_2")]
pub use cw_utils;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use derivative;
pub use sylvia_derive::{contract, entry_points, interface};
pub use {
    cosmwasm_schema as cw_schema, cosmwasm_std as cw_std, schemars, serde,
    serde_cw_value as serde_value, serde_json_wasm as serde_json,
};
