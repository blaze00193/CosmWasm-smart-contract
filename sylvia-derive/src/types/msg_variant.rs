use crate::crate_module;
use crate::fold::StripSelfPath;
use crate::parser::attributes::VariantAttrForwarding;
use crate::parser::check_generics::{CheckGenerics, GetPath};
use crate::parser::variant_descs::VariantDescs;
use crate::parser::{process_fields, MsgAttr, MsgType};
use crate::utils::{extract_return_type, filter_wheres, SvCasing};
use convert_case::{Case, Casing};
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use syn::fold::Fold;
use syn::visit::Visit;
use syn::{parse_quote, Ident, Signature, Type, WhereClause, WherePredicate};

use super::msg_field::MsgField;

/// Representation of whole message variant
#[derive(Debug)]
pub struct MsgVariant<'a> {
    name: Ident,
    function_name: &'a Ident,
    fields: Vec<MsgField<'a>>,
    /// Type extracted only in case of `Query` and used in `cosmwasm_schema::QueryResponses`
    /// `returns` attribute.
    return_type: Option<Type>,
    msg_type: MsgType,
    attrs_to_forward: Vec<VariantAttrForwarding>,
}

impl<'a> MsgVariant<'a> {
    /// Creates new message variant from trait method
    pub fn new<Generic>(
        sig: &'a Signature,
        generics_checker: &mut CheckGenerics<Generic>,
        msg_attr: MsgAttr,
        attrs_to_forward: Vec<VariantAttrForwarding>,
    ) -> MsgVariant<'a>
    where
        Generic: GetPath + PartialEq,
    {
        let function_name = &sig.ident;

        let name = function_name.to_case(Case::UpperCamel);
        let fields = process_fields(sig, generics_checker);
        let msg_type = msg_attr.msg_type();

        let return_type = if let MsgAttr::Query { resp_type, .. } = msg_attr {
            match resp_type {
                Some(resp_type) => {
                    let resp_type = parse_quote! { #resp_type };
                    generics_checker.visit_type(&resp_type);
                    Some(resp_type)
                }
                None => {
                    let return_type = extract_return_type(&sig.output);
                    let stripped_return_type = StripSelfPath.fold_path(return_type.clone());
                    generics_checker.visit_path(&stripped_return_type);
                    Some(parse_quote! { #return_type })
                }
            }
        } else {
            None
        };

        Self {
            name,
            function_name,
            fields,
            return_type,
            msg_type,
            attrs_to_forward,
        }
    }

    /// Emits message variant
    pub fn emit(&self) -> TokenStream {
        let Self {
            name,
            fields,
            msg_type,
            return_type,
            attrs_to_forward,
            ..
        } = self;
        let fields = fields.iter().map(MsgField::emit);
        let returns_attribute = msg_type.emit_returns_attribute(return_type);
        let attrs_to_forward = attrs_to_forward.iter().map(|attr| &attr.attrs);

        quote! {
            #returns_attribute
            #( #[ #attrs_to_forward ] )*
            #name {
                #(#fields,)*
            }
        }
    }

    /// Emits match leg dispatching against this variant. Assumes enum variants are imported into the
    /// scope. Dispatching is performed by calling the function this variant is build from on the
    /// `contract` variable, with `ctx` as its first argument - both of them should be in scope.
    pub fn emit_dispatch_leg(&self) -> TokenStream {
        let Self {
            name,
            fields,
            function_name,
            msg_type,
            ..
        } = self;

        let args: Vec<_> = fields
            .iter()
            .zip(1..)
            .map(|(field, num)| Ident::new(&format!("field{}", num), field.name().span()))
            .collect();

        let fields = fields
            .iter()
            .map(MsgField::name)
            .zip(args.clone())
            .map(|(field, num_field)| quote!(#field : #num_field));

        let method_call = msg_type.emit_dispatch_leg(function_name, &args);

        quote! {
            #name {
                #(#fields,)*
            } => #method_call
        }
    }

    /// Emits variants constructors. Constructors names are variants names in snake_case.
    pub fn emit_variants_constructors(&self) -> TokenStream {
        let Self { name, fields, .. } = self;

        let method_name = name.to_case(Case::Snake);
        let parameters = fields.iter().map(MsgField::emit_method_field);
        let arguments = fields.iter().map(MsgField::name);

        quote! {
            pub fn #method_name( #(#parameters),*) -> Self {
                Self :: #name { #(#arguments),* }
            }
        }
    }

    pub fn as_fields_names(&self) -> Vec<&Ident> {
        self.fields.iter().map(MsgField::name).collect()
    }

    pub fn emit_method_field(&self) -> Vec<TokenStream> {
        self.fields
            .iter()
            .map(MsgField::emit_method_field)
            .collect()
    }

    pub fn name(&self) -> &Ident {
        &self.name
    }

    pub fn function_name(&self) -> &Ident {
        self.function_name
    }

    pub fn fields(&self) -> &Vec<MsgField> {
        &self.fields
    }

    pub fn msg_type(&self) -> &MsgType {
        &self.msg_type
    }

    pub fn return_type(&self) -> &Option<Type> {
        &self.return_type
    }
}

#[derive(Debug)]
pub struct MsgVariants<'a, Generic> {
    variants: Vec<MsgVariant<'a>>,
    used_generics: Vec<&'a Generic>,
    unused_generics: Vec<&'a Generic>,
    where_predicates: Vec<&'a WherePredicate>,
    msg_ty: MsgType,
}

impl<'a, Generic> MsgVariants<'a, Generic>
where
    Generic: GetPath + PartialEq + ToTokens,
{
    pub fn new(
        source: VariantDescs<'a>,
        msg_ty: MsgType,
        all_generics: &'a [&'a Generic],
        unfiltered_where_clause: &'a Option<WhereClause>,
    ) -> Self {
        let mut generics_checker = CheckGenerics::new(all_generics);
        let variants: Vec<_> = source
            .filter_map(|variant_desc| {
                let msg_attr: MsgAttr = variant_desc.attr_msg()?;
                let attrs_to_forward = variant_desc.attrs_to_forward();

                if msg_attr.msg_type() != msg_ty {
                    return None;
                }

                Some(MsgVariant::new(
                    variant_desc.into_sig(),
                    &mut generics_checker,
                    msg_attr,
                    attrs_to_forward,
                ))
            })
            .collect();

        let (used_generics, unused_generics) = generics_checker.used_unused();
        let where_predicates = filter_wheres(unfiltered_where_clause, all_generics, &used_generics);

        Self {
            variants,
            used_generics,
            unused_generics,
            where_predicates,
            msg_ty,
        }
    }

    pub fn where_clause(&self) -> Option<WhereClause> {
        let where_predicates = &self.where_predicates;
        if !where_predicates.is_empty() {
            Some(parse_quote! { where #(#where_predicates),* })
        } else {
            None
        }
    }

    pub fn variants(&self) -> impl Iterator<Item = &MsgVariant> {
        self.variants.iter()
    }

    pub fn used_generics(&self) -> &Vec<&'a Generic> {
        &self.used_generics
    }

    pub fn unused_generics(&self) -> &Vec<&'a Generic> {
        &self.unused_generics
    }

    pub fn msg_ty(&self) -> MsgType {
        self.msg_ty
    }

    pub fn emit_phantom_match_arm(&self) -> TokenStream {
        let sylvia = crate_module();
        let Self { used_generics, .. } = self;
        if used_generics.is_empty() {
            return quote! {};
        }
        quote! {
            _Phantom(_) => Err(#sylvia ::cw_std::StdError::generic_err("Phantom message should not be constructed.")).map_err(Into::into),
        }
    }

    pub fn emit_dispatch_legs(&self) -> impl Iterator<Item = TokenStream> + '_ {
        self.variants
            .iter()
            .map(|variant| variant.emit_dispatch_leg())
    }

    pub fn as_names_snake_cased(&self) -> Vec<String> {
        self.variants
            .iter()
            .map(|variant| variant.name.to_string().to_case(Case::Snake))
            .collect()
    }

    pub fn emit_constructors(&self) -> impl Iterator<Item = TokenStream> + '_ {
        self.variants
            .iter()
            .map(MsgVariant::emit_variants_constructors)
    }

    pub fn emit(&self) -> impl Iterator<Item = TokenStream> + '_ {
        self.variants.iter().map(MsgVariant::emit)
    }

    pub fn get_only_variant(&self) -> Option<&MsgVariant> {
        self.variants.first()
    }

    pub fn emit_phantom_variant(&self) -> TokenStream {
        let Self {
            msg_ty,
            used_generics,
            ..
        } = self;

        if used_generics.is_empty() {
            return quote! {};
        }

        let return_attr = match msg_ty {
            MsgType::Query => quote! { #[returns((#(#used_generics,)*))] },
            _ => quote! {},
        };

        quote! {
            #[serde(skip)]
            #return_attr
            _Phantom(std::marker::PhantomData<( #(#used_generics,)* )>),
        }
    }
}
