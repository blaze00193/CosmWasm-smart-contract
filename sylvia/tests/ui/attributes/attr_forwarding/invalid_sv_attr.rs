#![allow(unused_imports)]
use sylvia::cw_std::{Response, StdResult};
use sylvia::types::{ExecCtx, InstantiateCtx};

pub struct Contract;

#[sylvia::contract]
#[sv::msg_attr(exec,)]
#[sv::msg_attr(random_msg, PartialOrd)]
#[sv::msg_attr(exec PartialOrd)]
#[sv::msg_attr(PartialOrd)]
impl Contract {
    pub const fn new() -> Self {
        Self
    }

    #[sv::msg(instantiate)]
    fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
        Ok(Response::new())
    }

    #[sv::msg(exec)]
    fn exec(&self, _ctx: ExecCtx) -> StdResult<Response> {
        Ok(Response::new())
    }
}

fn main() {}
