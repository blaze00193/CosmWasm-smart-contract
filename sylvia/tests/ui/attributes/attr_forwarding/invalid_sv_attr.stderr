error: Expected attribute of the form: `#[sv::msg_attr(msg_type, attribute_to_forward)]`
 --> tests/ui/attributes/attr_forwarding/invalid_sv_attr.rs:8:16
  |
8 | #[sv::msg_attr(exec,)]
  |                ^^^^

error: Invalid message type, expected one of: `exec`, `query`, `instantiate`, `migrate`, `reply` or `sudo`.
 --> tests/ui/attributes/attr_forwarding/invalid_sv_attr.rs:9:16
  |
9 | #[sv::msg_attr(random_msg, PartialOrd)]
  |                ^^^^^^^^^^

error: Expected attribute of the form: `#[sv::msg_attr(msg_type, attribute_to_forward)]`
  --> tests/ui/attributes/attr_forwarding/invalid_sv_attr.rs:10:16
   |
10 | #[sv::msg_attr(exec PartialOrd)]
   |                ^^^^

error: Expected attribute of the form: `#[sv::msg_attr(msg_type, attribute_to_forward)]`
  --> tests/ui/attributes/attr_forwarding/invalid_sv_attr.rs:11:16
   |
11 | #[sv::msg_attr(PartialOrd)]
   |                ^^^^^^^^^^
